Sources/Basic/CacheableSequence.swift

```diff
@@ -0,0 +1,74 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// Wrapper for caching an arbitrary sequnce.
+public final class CacheableSequence<T: Sequence>: Sequence {
+    public typealias Element = T.Element
+    public typealias Iterator = CacheableSequenceIterator<T>
+
+    /// The list of consumed items.
+    fileprivate var items: [Element] = []
+    
+    /// An iterator on the underlying sequence, until complete.
+    fileprivate var it: T.Iterator?
+    
+    public init(_ sequence: T) {
+        self.it = sequence.makeIterator()
+    }
+    
+    public func makeIterator() -> Iterator {
+        return CacheableSequenceIterator(self)
```

Comment:
Can we not use any iterator?

Reply:
It is better not to, this has stronger type information... clients can erase it if need be. I think, maybe you had something else in mind?

Reply:
Not sure if iterator type is really needed. AnyIterator does retain the element type.